### PR TITLE
 i18n common components: logs

### PR DESCRIPTION
**Pull request by rebeccaalpert**
Addressed details page logs. 

This is a part of https://issues.redhat.com/projects/CONSOLE/issues/CONSOLE-2416.

### CODE REVIEW

**Comment by spadgett** [comment]
/assign @TheRealJon 
@TheRealJon Can you take a look since you're most familiar with logs?

File: frontend/public/components/utils/resource-log.jsx [hunk 1]
```diff
@@ -164,316 +170,290 @@ LogControls.propTypes = {
...
-      this.ws = new WSFactory(watchURL, wsOpts)
-        .onclose(this._onClose)
-        .onerror(this._onError)
-        .onmessage(this._onMessage)
-        .onopen(this._onOpen);
+  React.useEffect(() => {
+    const resourceStarted =
+      prevResourceStatus === LOG_SOURCE_WAITING && resourceStatus !== LOG_SOURCE_WAITING;
+    if (resourceStarted) {
+      restartStream();
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx:367]:
```suggestion
```

File: frontend/public/components/utils/resource-log.jsx [hunk 2]
```diff
@@ -164,316 +170,290 @@ LogControls.propTypes = {
...
     }
-  }
+  }, [prevResourceStatus, resourceStatus, restartStream]);
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx:369]:
```suggestion
```

**Review by TheRealJon — CHANGES_REQUESTED** [review]
If we are going to migrate this to a function component, we might as well migrate it to TypeScript as well.

Tested locally and this works. Eslint was complaining about some potential bugs, mainly with hook dependencies, so I made some suggestions to improve a few hooks and dependencies.

**Comment by TheRealJon** [comment]
@rebeccaalpert I didn't test these suggestions and work out any kinks they may cause, so take them with a grain of salt. Also, I realized after the fact that the websocket handlers actually could be declared inside the `wsInit` callback. Those handlers are not used anywhere else, so that may help narrow down the deps for that hook.

File: frontend/public/components/utils/resource-log.jsx [hunk 3]
```diff
@@ -164,316 +170,290 @@ LogControls.propTypes = {
...
 
   // Handler for websocket onopen event
-  _onOpen() {
-    this._buffer.clear();
-    this._updateStatus(STREAM_ACTIVE);
-  }
+  const onOpen = () => {
+    buffer.clear();
+    updateStatus(STREAM_ACTIVE);
+  };
```
[the remark this answers is not in the document]
Comment (TheRealJon) [reply frontend/public/components/utils/resource-log.jsx:236]: also `buffer.current.clear()`

File: frontend/public/components/utils/resource-log.jsx (hunk 2 above)
Comment (rebeccaalpert) [reply frontend/public/components/utils/resource-log.jsx:369]: Did you mean to suggest something here? It looks identical.

File: frontend/public/components/utils/resource-log.jsx (hunk 1 above)
Comment (rebeccaalpert) [reply frontend/public/components/utils/resource-log.jsx:367]: Did you mean to suggest something here? It looks identical.

**Comment by rebeccaalpert** [comment]
Thanks! I incorporated the first round and I'm testing; I'll look into declaring the handlers in wsInit afterwards.

**Comment by TheRealJon** [comment]
Hi @rebeccaalpert I tested my suggested changes a bit last week and there some issues with render loops caused by the new hook dependencies. If you can't work those issues out easily, I would just make whatever improvements you can without breaking your original implementation. It would be nice if we could update this component to be more in line with our current conventions, but this component is also a bit complicated, so maybe we can fix more of this in a follow on.

File: frontend/public/components/utils/resource-log.jsx [hunk 4]
```diff
@@ -164,316 +170,290 @@ LogControls.propTypes = {
...
-
+  };
   // Handler for websocket onclose event
-  _onClose() {
-    this.setState({ status: STREAM_EOF });
-  }
-
+  const onClose = () => {
+    setStatus(STREAM_EOF);
+  };
```
[the remark this answers is not in the document]
Comment (rebeccaalpert) [reply frontend/public/components/utils/resource-log.jsx:204]: Skipping in lieu of moving into wsInit hook.

File: frontend/public/components/utils/resource-log.jsx [hunk 5]
```diff
@@ -164,316 +170,290 @@ LogControls.propTypes = {
...
   // Handler for websocket onerror event
-  _onError() {
-    this.setState({
-      error: true,
-    });
-  }
-
+  const onError = () => {
+    setError(true);
+  };
```
[the remark this answers is not in the document]
Comment (rebeccaalpert) [reply frontend/public/components/utils/resource-log.jsx:208]: Skipping in lieu of moving into wsInit hook.

File: frontend/public/components/utils/resource-log.jsx [hunk 6]
```diff
@@ -164,316 +170,290 @@ LogControls.propTypes = {
...
   // Handler for websocket onmessage event
-  _onMessage(msg) {
-    const { linesBehind, status } = this.state;
+  const onMessage = (msg) => {
```
[the remark this answers is not in the document]
Comment (rebeccaalpert) [reply frontend/public/components/utils/resource-log.jsx:210]: Skipping in lieu of moving into wsInit hook.

File: frontend/public/components/utils/resource-log.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (rebeccaalpert) [reply frontend/public/components/utils/resource-log.jsx:236]: Skipping useCallback in lieu of moving into wsInit hook.

**Review by TheRealJon — APPROVED** [review]
/lgtm
/approve

**Comment by rebeccaalpert** [comment]
Hey @TheRealJon  - I narrowed down the issue to a useCallback and a useEffect. It seems to work ok on my end now. I also converted the component to TypeScript. Can you take a look and let me know if you have additional feedback? Thanks!

File: frontend/public/components/utils/resource-log.jsx [hunk 7]
```diff
@@ -164,316 +170,290 @@ LogControls.propTypes = {
...
-      });
+      const linesAdded = buffer.ingest(text);
+      setLinesBehind(status === STREAM_PAUSED ? linesBehind + linesAdded : linesBehind);
+      setLines([...buffer.getLines(), buffer.getTail()]);
+      setHasTruncated(buffer.getHasTruncated());
+    }
+  };
+
+  // Updates log status
+  const updateStatus = (newStatus) => {
```
[the remark this answers is not in the document]
Comment (rebeccaalpert) [reply frontend/public/components/utils/resource-log.jsx]: This seemed to lead to a loop, so I left it out.

File: frontend/public/components/utils/resource-log.jsx [hunk 8]
```diff
@@ -164,316 +170,290 @@ LogControls.propTypes = {
...
+        prevResourceStatus === LOG_SOURCE_RESTARTING &&
+        resourceStatus !== LOG_SOURCE_RESTARTING
+      ) {
+        newState.stale = true;
+      }
+      setPrevResourceStatus(resourceStatus);
+      setResourceStatus(newState.resourceStatus);
+      newState.stale && setStale(newState.stale);
+    }
+  }, [props.resourceStatus, prevResourceStatus, resourceStatus]);
```
[the remark this answers is not in the document]
Comment (rebeccaalpert) [reply frontend/public/components/utils/resource-log.jsx]: This left out handling the initial component state (it was loading forever). I updated the original useEffect hooks to use props.resourceStatus and resourceStatus as outlined above.

File: frontend/public/components/utils/resource-log.tsx [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+  [STREAM_EOF]: 'logs~Log stream ended.',
+  // t('logs~Loading log...')
+  [STREAM_LOADING]: 'logs~Loading log...',
+  // t('logs~Log stream paused.')
+  [STREAM_PAUSED]: 'logs~Log stream paused.',
+  // t('logs~Log streaming...')
+  [STREAM_ACTIVE]: 'logs~Log streaming...',
+};
+
+const replaceVariables = (template, values) => {
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.tsx:46]: type defs

**Review by spadgett — APPROVED** [review]
/lgtm
/hold

**Comment by spadgett** [comment]
@rebeccaalpert There's a merge conflict here. We should squash the intermediate commits (feel free to leave @TheRealJon's though).

/lgtm cancel

**Review by spadgett — APPROVED** [review]
/lgtm
/hold

**Comment by spadgett** [comment]
This is ready for test.

/assign @yapei @ahardin-rh @sferich888 
for approvals

**Pull request merged**